### ISSUE TITLE
error: Error in src/extproc/extproc_spawner.cc at line 219:

### ISSUE THREAD

**Issue by fluxxu** [rethinkdb/rethinkdb#5572]
```
Server ready, "5632ee0b598d_4dk" 70f68380-f7d9-4d56-b0e6-2a9d000fc9d8
Running rethinkdb 2.2.5~0jessie (GCC 4.9.2)...
Running on Linux 4.1.5-x86_64-linode61 x86_64
Loading data from directory /data/rethinkdb_data
warn: Cache size does not leave much memory for server and query overhead (available memory: 629 MB).
warn: Cache size is very low and may impact performance.
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on addresses: 127.0.0.1, 172.17.0.3, ::1, fe80::42:acff:fe11:3%60
Server ready, "5632ee0b598d_4dk" 70f68380-f7d9-4d56-b0e6-2a9d000fc9d8
warn: Problem when checking for new versions of RethinkDB: HTTP request to update.rethinkdb.com failed.
Version: rethinkdb 2.2.5~0jessie (GCC 4.9.2)
error: Error in src/extproc/extproc_spawner.cc at line 219:
error: Guarantee failed: [res == 0]  (errno 32 - Broken pipe) could not send socket file descriptor to worker process
error: Backtrace:
error: Fri Mar 25 21:50:21 2016

       1 [0xa8b7f0]: backtrace_t::backtrace_t() at 0xa8b7f0 (rethinkdb)
       2 [0xa8bb69]: format_backtrace(bool) at 0xa8bb69 (rethinkdb)
       3 [0xd0135d]: report_fatal_error(char const*, int, char const*, ...) at 0xd0135d (rethinkdb)
       4 [0xcefc2e]: extproc_spawner_t::spawn(object_buffer_t<socket_stream_t>*, int*) at 0xcefc2e (rethinkdb)
       5 [0xcf55f8]: extproc_worker_t::acquired(signal_t*) at 0xcf55f8 (rethinkdb)
       6 [0xceedb3]: extproc_job_t::extproc_job_t(extproc_pool_t*, bool (*)(read_stream_t*, write_stream_t*), signal_t*) at 0xceedb3 (rethinkdb)
       7 [0xcf46a2]: http_runner_t::http(http_opts_t const&, http_result_t*, signal_t*) at 0xcf46a2 (rethinkdb)
       8 [0x86de6c]: ql::dispatch_http(ql::env_t*, http_opts_t const&, http_runner_t*, http_result_t*, ql::bt_rcheckable_t const*) at 0x86de6c (rethinkdb)
       9 [0xbfd9bc]: version_checker_t::do_check(bool, auto_drainer_t::lock_t) at 0xbfd9bc (rethinkdb)
       10 [0xbfeb84]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (version_checker_t::*)(bool, auto_drainer_t::lock_t)> (version_checker_t*, bool, auto_drainer_t::lock_t)> >::run_action() at 0xbfeb84 (rethinkdb)
       11 [0x99c958]: coro_t::run() at 0x99c958 (rethinkdb)
error: Exiting.
```

Is this because the server has only 1GB of RAM (512MB swap)?
Thanks!

**Comment by danielmewes** [comment]
Thanks for reporting this @fluxxu . It's possible that this is because of low memory, but it might also be a bug of some sort.

@Tryneus do you know anything about this error?

**Comment by danielmewes** [comment]
We have two other reports of this (or a similar) extproc spawner issue: https://github.com/rethinkdb/rethinkdb/issues/5837
https://github.com/rethinkdb/rethinkdb/issues/5912

@Tryneus could we at least change the extproc spawner to not crash when encountering this issue?

**Comment by Tryneus** [comment]
@danielmewes, taking a quick look it shouldn't be hard to cause an error on a given query if an error occurs when trying to launch a worker process.  Honestly in extproc code, we should probably only guarantee at startup, and anything run during normal operations should just fail the current query (and probably log something).

**Comment by wspeirs** [comment]
I'm seeing the same crashes because of the version checker:

```
warn: Problem when checking for new versions of RethinkDB: HTTP request to update.rethinkdb.com failed.
error: Error in src/extproc/extproc_spawner.cc at line 221:
error: Guarantee failed: [res == 0]  (errno 32 - Broken pipe) could not send socket file descriptor to worker process
error: Backtrace:
error: Fri Jul  8 16:19:02 2016

1 [0xabb110]: backtrace_t::backtrace_t() at ??:?
2 [0xabb4a3]: format_backtrace(bool) at ??:?
3 [0xd3e78d]: report_fatal_error(char const*, int, char const*, ...) at ??:?
4 [0xd2e3fc]: extproc_spawner_t::spawn(object_buffer_t<socket_stream_t>*, int*) at ??:?
5 [0xd34158]: extproc_worker_t::acquired(signal_t*) at ??:?
6 [0xd2d453]: extproc_job_t::extproc_job_t(extproc_pool_t*, bool (*)(read_stream_t*, write_stream_t*), signal_t*) at ??:?
7 [0xd32ee5]: http_runner_t::http(http_opts_t const&, http_result_t*, signal_t*) at ??:?
8 [0x8bb758]: ql::dispatch_http(ql::env_t*, http_opts_t const&, http_runner_t*, http_result_t*, ql::bt_rcheckable_t const*) at ??:?
9 [0xc20174]: version_checker_t::do_check(bool, auto_drainer_t::lock_t) at ??:?
10 [0xc20d42]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (version_checker_t::*)(bool, auto_drainer_t::lock_t)> (version_checker_t*, bool, auto_drainer_t::lock_t)> >::run_action() at ??:?
11 [0x9dab27]: coro_t::run() at ??:?

error: Exiting.
```

I have a LOT (570M documents) in a 3-node cluster. Each node has 256GB of memory, but memory is pretty tight: 92% of cache used.

**Comment by danielmewes** [comment]
@wspeirs If it's just the update check, I recommend disabling it as a work-around like described here: https://github.com/rethinkdb/rethinkdb/issues/5837#issuecomment-225042140

**Comment by 1lann** [comment]
I just wanted to say that I had the same issue on `rethinkdb 2.3.4~0xenial (GCC 5.3.1)`, and similarly to @fluxxu, the error occurs when checking for updates and I am too using very little memory with RethinkDB (512 MB of RAM with a cache size of 200 MB) in a 2 node cluster (I'm not using the database for anything serious) where the cache usage is consistently > 90%.

Log:
```
2016-07-18T07:47:31.839636103 777605.340420s warn: Problem when checking for new versions of RethinkDB: HTTP request to update.rethinkdb.com failed.
2016-07-19T07:47:31.829772777 864005.330558s error: Error in src/extproc/extproc_spawner.cc at line 221:
2016-07-19T07:47:31.831181907 864005.331967s error: Guarantee failed: [res == 0]  (errno 32 - Broken pipe) could not send socket file descriptor to worker process
2016-07-19T07:47:31.831249866 864005.332034s error: Backtrace:
2016-07-19T07:47:32.105239019 864005.606026s error: Tue Jul 19 07:47:31 2016\n\n1 [0xb304fa]: backtrace_t::backtrace_t() at ??:?\n2 [0xb309da]: format_backtrace[abi:cxx11](bool) at ??:?\n3 [0xdf905c]: report_fatal_error(char const*, int, char const*, ...) at ??:?\n4 [0xde67b6]: extproc_spawner_t::spawn(object_buffer_t<socket_stream_t>*, int*) at ??:?\n5 [0xdec218]: extproc_worker_t::acquired(signal_t*) at ??:?\n6 [0xde5863]: extproc_job_t::extproc_job_t(extproc_pool_t*, bool (*)(read_stream_t*, write_stream_t*), signal_t*) at ??:?\n7 [0xdeae03]: http_runner_t::http(http_opts_t const&, http_result_t*, signal_t*) at ??:?\n8 [0x8f9e95]: ql::dispatch_http(ql::env_t*, http_opts_t const&, http_runner_t*, http_result_t*, ql::bt_rcheckable_t const*) at ??:?\n9 [0xcbe8f0]: version_checker_t::do_check(bool, auto_drainer_t::lock_t) at ??:?\n10 [0xcbfbc4]: callable_action_instance_t<std::_Bind<std::_Mem_fn<void (version_checker_t::*)(bool, auto_drainer_t::lock_t)> (version_checker_t*, bool, auto_drainer_t::lock_t)> >::run_action() at ??:?\n11 [0xa3a788]: coro_t::run() at ??:?
2016-07-19T07:47:32.105361209 864005.606145s error: Exiting.
```

**Comment by danielmewes** [comment]
@1lann Can you check `dmesg` for whether there's any indication of the out-of-memory handler has terminated a `rethinkdb` process? I wonder if this crash will happen if the separate process that we spawn for performing the update check gets terminated because it's out of memory.

**Comment by 1lann** [comment]
Sorry for taking so long to respond, I forgot to earlier! There was nothing logged to dmesg, and unfortunately syslog had cycled over and I don't have logs from that long ago. However, I did remember seeing something in the syslog regarding RethinkDB when it quit, although it wasn't the OOM killer, but it was a single line entry which I forgot the exact contents of (was quite terse, I don't think it was that human readable). I believe it contained an exit code, and I dismissed it as just logging RethinkDB quitting because it was a service running on startup by init.d, however it might have be caused by an out of memory when the child process was created.

**Comment by donpdonp** [comment]
I'm seeing this too in the latest rethinkdb docker image running on coreos 1068.

```
Aug 19 03:22:58 db1 docker[896]: Version: rethinkdb 2.3.4~0jessie (GCC 4.9.2)
Aug 19 03:22:59 db1 docker[896]: error: Error in src/extproc/extproc_spawner.cc at line 221:
Aug 19 03:22:59 db1 docker[896]: error: Guarantee failed: [res == 0]  (errno 32 - Broken pipe) could n
ot send socket file descriptor to worker process
Aug 19 03:22:59 db1 docker[896]: error: Backtrace:
Aug 19 03:22:59 db1 docker[896]: error: Fri Aug 19 03:22:59 2016
Aug 19 03:22:59 db1 docker[896]:        
Aug 19 03:22:59 db1 docker[896]:        1 [0xae5060]: backtrace_t::backtrace_t() at 0xae5060 (rethinkd
b)
Aug 19 03:22:59 db1 docker[896]:        2 [0xae53d9]: format_backtrace(bool) at 0xae53d9 (rethinkdb)
Aug 19 03:22:59 db1 docker[896]:        3 [0xd9abf3]: report_fatal_error(char const*, int, char const*
, ...) at 0xd9abf3 (rethinkdb)
Aug 19 03:22:59 db1 docker[896]:        4 [0xd8979e]: extproc_spawner_t::spawn(object_buffer_t<socket_
stream_t>*, int*) at 0xd8979e (rethinkdb)
...
```

**Comment by danielmewes** [comment]
@donpdonp How much RAM is available to the container?

**Comment by donpdonp** [comment]
@danielmewes 1gb is allocated to coreos.

**Comment by srh** [comment]
Anybody who finds this issue: You can lower the chance of seeing this bug if you pass `--no-update-check`.

**Comment by srh** [comment]
This has been fixed in v2.3.x, v2.4, and next.
